File: stacks/node/base/code/testing.tpl.js
/*
File: testing.tpl.js
Purpose: Testing utilities for Jest/Vitest
Generated for: {{PROJECT_NAME}}
*/

const { beforeAll, afterAll, beforeEach, afterEach, describe, it, expect } = require('@jest/globals');

// Test data factory
const TestDataFactory = {
    user(overrides = {}) {
        return {
            email: 'test@example.com',
            username: 'testuser',
            password: 'Password123!',
            ...overrides,
        };
    },

    post(overrides = {}) {
        return {
            title: 'Test Post',
            content: 'Test content',
            ...overrides,
        };
    },

    // Generate unique email
    uniqueEmail() {
        return `test-${Date.now()}-${Math.random().toString(36).slice(2)}@example.com`;
    },

    // Generate unique username
    uniqueUsername() {
        return `user-${Date.now()}-${Math.random().toString(36).slice(2)}`;
    },
};

// Mock response helper
class MockResponse {
    constructor(data, status = 200) {
        this.data = data;
        this.status = status;
        this.ok = status >= 200 && status < 300;
    }

    json() {
        return Promise.resolve(this.data);
    }

    text() {
        return Promise.resolve(JSON.stringify(this.data));
    }
}

// Mock fetch helper
function mockFetch(responses = {}) {
    return jest.fn((url, options = {}) => {
        const method = options.method || 'GET';
        const key = `${method} ${url}`;

        if (responses[key]) {
            const response = responses[key];
            return Promise.resolve(new MockResponse(response.data, response.status));
        }

        return Promise.resolve(new MockResponse({ error: 'Not found' }, 404));
    });
}

// Database test helpers
class TestDatabase {
    constructor(prisma) {
        this.prisma = prisma;
    }

    async reset() {
        // Delete in correct order to respect foreign keys
        const tablenames = await this.prisma.$queryRaw`
      SELECT tablename FROM pg_tables WHERE schemaname = 'public'
    `;

        for (const { tablename } of tablenames) {
            if (tablename !== '_prisma_migrations') {
                await this.prisma.$executeRawUnsafe(`TRUNCATE TABLE "${tablename}" CASCADE`);
            }
        }
    }

    async seed(data = {}) {
        // Seed test data
        if (data.users) {
            await this.prisma.user.createMany({ data: data.users });
        }
        // Add more seeders as needed
    }
}

// Request helper for API testing
class TestClient {
    constructor(baseUrl = 'http://localhost:3000') {
        this.baseUrl = baseUrl;
        this.headers = {};
    }

    setAuth(token) {
        this.headers['Authorization'] = `Bearer ${token}`;
    }

    clearAuth() {
        delete this.headers['Authorization'];
    }

    async request(method, path, { body, query, headers = {} } = {}) {
        const url = new URL(path, this.baseUrl);

        if (query) {
            Object.entries(query).forEach(([key, value]) => {
                url.searchParams.append(key, value);
            });
        }

        const response = await fetch(url.toString(), {
            method,
            headers: {
                'Content-Type': 'application/json',
                ...this.headers,
                ...headers,
            },
            body: body ? JSON.stringify(body) : undefined,
        });

        const data = await response.json().catch(() => null);

        return {
            status: response.status,
            ok: response.ok,
            data,
            headers: Object.fromEntries(response.headers.entries()),
        };
    }

    get(path, options) {
        return this.request('GET', path, options);
    }

    post(path, body, options) {
        return this.request('POST', path, { ...options, body });
    }

    put(path, body, options) {
        return this.request('PUT', path, { ...options, body });
    }

    patch(path, body, options) {
        return this.request('PATCH', path, { ...options, body });
    }

    delete(path, options) {
        return this.request('DELETE', path, options);
    }
}

// Assertion helpers
const assertions = {
    assertOk(response, expectedStatus = 200) {
        expect(response.status).toBe(expectedStatus);
    },

    assertError(response, expectedStatus, errorCode) {
        expect(response.status).toBe(expectedStatus);
        if (errorCode) {
            expect(response.data?.error?.code).toBe(errorCode);
        }
    },

    assertContainsKeys(obj, keys) {
        keys.forEach((key) => {
            expect(obj).toHaveProperty(key);
        });
    },

    assertArrayLength(arr, length) {
        expect(arr).toHaveLength(length);
    },
};

// Timeout helpers
function wait(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
}

async function waitFor(condition, { timeout = 5000, interval = 100 } = {}) {
    const start = Date.now();
    while (Date.now() - start < timeout) {
        if (await condition()) {
            return true;
        }
        await wait(interval);
    }
    throw new Error(`Condition not met within ${timeout}ms`);
}

// Test context helper
function createTestContext() {
    const context = {
        data: {},
        cleanup: [],
    };

    return {
        set(key, value) {
            context.data[key] = value;
        },
        get(key) {
            return context.data[key];
        },
        onCleanup(fn) {
            context.cleanup.push(fn);
        },
        async cleanup() {
            for (const fn of context.cleanup.reverse()) {
                await fn();
            }
            context.data = {};
            context.cleanup = [];
        },
    };
}

module.exports = {
    TestDataFactory,
    MockResponse,
    mockFetch,
    TestDatabase,
    TestClient,
    assertions,
    wait,
    waitFor,
    createTestContext,
};

// Usage:
// const { TestDataFactory, TestClient, assertions } = require('./testing');
//
// describe('User API', () => {
//   const client = new TestClient();
//
//   it('should create user', async () => {
//     const response = await client.post('/users', TestDataFactory.user());
//     assertions.assertOk(response, 201);
//     assertions.assertContainsKeys(response.data, ['id', 'email']);
//   });
// });

File: stacks/python/base/code/testing.tpl.py
"""
File: testing.tpl.py
Purpose: Testing utilities and fixtures for pytest
Generated for: {{PROJECT_NAME}}
"""

import asyncio
from typing import AsyncGenerator, Generator, Any
from contextlib import asynccontextmanager
import pytest
from httpx import AsyncClient, ASGITransport
from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine, async_sessionmaker
from sqlalchemy.pool import StaticPool


# Test database URL (in-memory SQLite for tests)
TEST_DATABASE_URL = "sqlite+aiosqlite:///:memory:"


@pytest.fixture(scope="session")
def event_loop() -> Generator:
    """Create event loop for async tests"""
    loop = asyncio.get_event_loop_policy().new_event_loop()
    yield loop
    loop.close()


@pytest.fixture(scope="session")
async def test_engine():
    """Create test database engine"""
    engine = create_async_engine(
        TEST_DATABASE_URL,
        connect_args={"check_same_thread": False},
        poolclass=StaticPool,
    )
    yield engine
    await engine.dispose()


@pytest.fixture(scope="function")
async def db_session(test_engine) -> AsyncGenerator[AsyncSession, None]:
    """Get a database session for each test"""
    # Import your Base model here
    # from models import Base
    # async with test_engine.begin() as conn:
    #     await conn.run_sync(Base.metadata.create_all)

    session_maker = async_sessionmaker(
        test_engine,
        class_=AsyncSession,
        expire_on_commit=False,
    )

    async with session_maker() as session:
        yield session
        await session.rollback()


@pytest.fixture
async def client(app) -> AsyncGenerator[AsyncClient, None]:
    """Get async HTTP client for testing FastAPI app"""
    async with AsyncClient(
        transport=ASGITransport(app=app),
        base_url="http://test",
    ) as client:
        yield client


@pytest.fixture
def auth_headers():
    """Get authentication headers for protected endpoints"""
    def _headers(token: str = "test-token") -> dict:
        return {"Authorization": f"Bearer {token}"}
    return _headers


# Test utilities
class TestDataFactory:
    """Factory for creating test data"""

    @staticmethod
    def user_data(
        email: str = "test@example.com",
        username: str = "testuser",
        password: str = "Password123!",
        **kwargs,
    ) -> dict:
        return {
            "email": email,
            "username": username,
            "password": password,
            **kwargs,
        }

    @staticmethod
    def post_data(
        title: str = "Test Post",
        content: str = "Test content",
        **kwargs,
    ) -> dict:
        return {
            "title": title,
            "content": content,
            **kwargs,
        }


# Async context managers for test setup/teardown
@asynccontextmanager
async def create_test_user(db_session: AsyncSession, **kwargs):
    """Create a test user and clean up after"""
    # from models import User
    # user = User(**TestDataFactory.user_data(**kwargs))
    # db_session.add(user)
    # await db_session.commit()
    # await db_session.refresh(user)
    # try:
    #     yield user
    # finally:
    #     await db_session.delete(user)
    #     await db_session.commit()
    yield None  # Placeholder


# Assertion helpers
def assert_response_ok(response, expected_status: int = 200):
    """Assert response is successful"""
    assert response.status_code == expected_status, (
        f"Expected {expected_status}, got {response.status_code}: {response.text}"
    )


def assert_response_error(response, expected_status: int, error_code: str = None):
    """Assert response is an error with expected status"""
    assert response.status_code == expected_status
    if error_code:
        data = response.json()
        assert data.get("error", {}).get("code") == error_code


def assert_contains_keys(data: dict, keys: list):
    """Assert dict contains all expected keys"""
    for key in keys:
        assert key in data, f"Missing key: {key}"


# Mock utilities
class MockResponse:
    """Mock HTTP response for testing"""

    def __init__(
        self,
        json_data: Any = None,
        status_code: int = 200,
        text: str = "",
    ):
        self._json_data = json_data
        self.status_code = status_code
        self.text = text or str(json_data)

    def json(self):
        return self._json_data

    def raise_for_status(self):
        if self.status_code >= 400:
            raise Exception(f"HTTP {self.status_code}")


# Usage in tests:
# @pytest.mark.asyncio
# async def test_create_user(client: AsyncClient, db_session: AsyncSession):
#     response = await client.post("/users", json=TestDataFactory.user_data())
#     assert_response_ok(response, 201)
#     assert_contains_keys(response.json(), ["id", "email", "username"])
